src/NuGet.Core/NuGet.Protocol/Resources/ServiceIndexResourceV3.cs

```diff
@@ -105,7 +108,22 @@ private IReadOnlyList<ServiceIndexEntry> GetBestVersionMatchForType(NuGetVersion
             else
             {
                 // Find all entries with the same version.
-                return entries.Where(e => e.ClientVersion == bestMatch.ClientVersion).ToList();
+                List<ServiceIndexEntry> matchingEntries = new List<ServiceIndexEntry>();
+
+                foreach (var entry in entries)
+                {
+                    if (entry.ClientVersion == bestMatch.ClientVersion)
+                    {
+                        if (entry.Uri.Scheme == Uri.UriSchemeHttp && entry.Uri.Scheme != Uri.UriSchemeHttps && !_allowInsecureConnections)
```

Comment:
I don't know if the deep link will work, since the comment is resolved in the other PR: https://github.com/NuGet/NuGet.Client/pull/5760#discussion_r1637292706

But there I suggested that rather than checking every single time the same resource is requested over and over, we can check once during initialization. But the trade off is that it means any service index entry that is unused, but pointing to an insecure HTTP URL, will prevent NuGet from working.

Reply:
After reviewing the issue, I see no major problem to performing the check each time a resource is requested. The current implementation includes a loop that goes through each service entry to locate the desired version. The http check is integrated within this loop, thus the performance impact is minimal, maintaining an O(n) complexity both before and after the change. In addition, this approach doesn't restrict users from accessing a secure resource just because another listed resource may be insecure. It's possible that the source provider is in the process of upgrading their service entries to HTTPS. As long as the insecure resource isn't being utilized, I don't believe we should block